src/lazy/binary/raw/v1_1/type_descriptor.rs

```diff
@@ -0,0 +1,141 @@
+/// For more information, consult the
+/// [Typed Value Formats](https://amazon-ion.github.io/ion-docs/docs/binary.html#typed-value-formats)
+/// section of the binary Ion spec.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct TypeDescriptor {
+    pub ion_type_code: IonTypeCode,
+    pub ion_type: Option<IonType>,
+    pub length_code: u8,
+}
+
+/// A statically defined array of TypeDescriptor that allows a binary reader to map a given
+/// byte (`u8`) to a `TypeDescriptor` without having to perform any masking or bitshift operations.
+pub(crate) static ION_1_1_TYPE_DESCRIPTORS: &[TypeDescriptor; 256] = &init_type_descriptor_cache();
+
+const DEFAULT_HEADER: TypeDescriptor = TypeDescriptor {
+    ion_type_code: IonTypeCode::Nop,
+    ion_type: None,
+    length_code: 0,
+};
+
+pub(crate) const fn init_type_descriptor_cache() -> [TypeDescriptor; 256] {
+    let mut jump_table = [DEFAULT_HEADER; 256];
+    let mut index: usize = 0;
+    while index < 256 {
+        let byte = index as u8;
+        jump_table[index] = TypeDescriptor::from_byte(byte);
+        index += 1;
+    }
+    jump_table
+}
+
+impl TypeDescriptor {
+    /// Attempts to parse the provided byte. If the type code is unrecognized or the
+    /// type code + length code combination is illegal, an error will be returned.
+    pub const fn from_byte(byte: u8) -> TypeDescriptor {
+        let (high_nibble, low_nibble) = (byte >> 4, byte & 0x0F);
+        use IonTypeCode::*;
+
+        let ion_type_code = match (high_nibble, low_nibble) {
+            (0xE, 0x0) => IonVersionMarker,
+            (0xE, 0xA) => NullNull,
+            (0xE, 0xC..=0xD) => Nop,
+            _ => Boolean, // Temporary, until everything is implemented to satisfy the LUT.
+        };
+        let ion_type = match ion_type_code {
+            NullNull => Some(IonType::Null),
+            Nop => None,
+            IonVersionMarker => None,
+            Boolean => Some(IonType::Bool),
+            _ => panic!("the provided ion type code is either not implemented, or invalid"),
+        };
+        TypeDescriptor {
+            ion_type,
+            ion_type_code,
+            length_code: low_nibble,
+        }
+    }
+
+    pub fn is_null(&self) -> bool {
+        self.ion_type_code == IonTypeCode::NullNull || self.ion_type_code == IonTypeCode::TypedNull
+    }
+
+    pub fn is_nop(&self) -> bool {
+        self.ion_type_code == IonTypeCode::Nop
+    }
+
+    pub fn is_ivm_start(&self) -> bool {
+        self.ion_type_code == IonTypeCode::IonVersionMarker
+    }
+
+    pub fn is_annotation_wrapper(&self) -> bool {
+        false
+    }
+
+    #[inline]
+    pub fn to_header(self) -> Option<Header> {
+        let ion_type = self.ion_type?;
+        let header = Header {
+            ion_type,
+            ion_type_code: self.ion_type_code,
+            length_code: self.length_code,
+        };
+        Some(header)
+    }
+}
+
+pub enum LengthType {
+    InHeader(usize),
```

Comment:
Often the phrase "header" includes the trailing length when present. Maybe:
```suggestion
    InOpcode(usize),
```
?

Reply:
The size in the opcode is never more than 15, right? If so, we can use `u8`.